### ISSUE TITLE
FTPClient should support deleting files.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3491]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tarjei reported|
|-|-|
|Trac ID|trac#3491|
|Type|enhancement|
|Created|2008-10-24 10:57:32Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-delete-3491-2|

The following patch adds support for deleting files in the FTP client. 

CC'ing spiv since he has reviewed my other patch to the ftpclient. 

Attachments:

* [ftp-delete.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02d/02d666af5843ace17fb3610658e88c60c2aa067c/bc870ac640a676da0397365e7212b9cded28ec15.diff) (3209 bytes) - added by tarjei on 2008-10-24 10:57:42Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3491 3491
type__enhancement enhancement
reporter__tarjei tarjei
priority__normal normal
milestone__ 
branch__branches_ftp_delete_3491_2 branches/ftp-delete-3491-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1224845852000000 1224845852000000
changetime__1225390059000000 1225390059000000
version__None None
owner__ 
cc__spiv cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="tarjei, exarkun's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|tarjei, exarkun commented|
|-|-|

(In [25074]) Branching to 'ftp-delete-3491'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [25075]) apply ftp-delete.diff

refs [#3491](https://github.com/twisted/twisted/issues/3491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Well, that was fun.

I made some further changes:

  1. `changeDirectory` is now deprecated.  The only difference it had from `cwd` is that it would incorrectly consider 2xx responses other than 250 to be error responses.
  1. The heap of code which used to be in the new `removeFile` method to massage the response from the server is gone.  It wasn't necessary.
  1. tweaked some docstrings and added some more tests

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

1.the @since tag is missing
 1. there's a typo, use explanations instead of explenations
 1. the items in 'Possible return codes:' should be listed using a '-'
 1. the '250' line is missing a ':', as used in the other list items

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25104]) add @since markup

refs [#3491](https://github.com/twisted/twisted/issues/3491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Fixed 1, I think 2-4 were based on a review of r25075, but the branch includes some more revisions after that which already fixed those issues.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix set owner to @exarkun|
|-|-|

I'm not sure why you changed the behavior of changeDirectory. Was it completely broken before? There aren't any new unit tests for the change in behavior, only for the fact that it's now deprecated.


The delete support looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @radix|
|-|-|

I didn't change any behavior of `changeDirectory` except to deprecate it.  All the code I deleted was redundant and stupid.  All of the code necessary to implement the behavior it provided is still present, in the new shortened `cbResult` method.  Since the behavior is still stupid, `changeDirectory` is deprecated now.  All it does is what `cwd` does except it considers some success responses (3xx) to be failures instead.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to @exarkun|
|-|-|

Ok that's a good enough explanation, +1

**Comment by twisted-trac** [comment]
|<img alt="tarjei, exarkun's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|tarjei, exarkun commented|
|-|-|

(In [25195]) Branching to 'ftp-delete-3491-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [25197]) Avoid garbage collector interaction

refs [#3491](https://github.com/twisted/twisted/issues/3491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @radix|
|-|-|

The new change directory test had the same bug as one of the make directory tests from [#3335](https://github.com/twisted/twisted/issues/3335).

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix set owner to @exarkun|
|-|-|

Ok, I guess that looks good. I guess I will use -u from now on. :(

maybe we should have a build slave that does -u=100!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [25201]) Merge ftp-delete-3491-2

Author: tarjei, exarkun
Reviewer: thijs, radix
Fixes: [#3491](https://github.com/twisted/twisted/issues/3491)

Add `FTPClient.removeFile` method for using the file deletion feature of FTP.

Also deprecate `FTPClient.changeDirectory` which added incorrect response
handling on top of `FTPClient.cwd`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
